[BUGFIX] Databricks SQL tables with special characters

fredwang1012 opened the pull request:
- [x] Description of PR changes above includes a link to [an existing GitHub issue](https://github.com/great-expectations/great_expectations/issues)
- [x] PR title is prefixed with one of: [BUGFIX], [FEATURE], [DOCS], [MAINTENANCE], [CONTRIB], [MINORBUMP]
- [x] Code is linted - run `invoke lint` (uses `ruff format` + `ruff check`)
- [x] Appropriate tests and docs have been updated

For more information about contributing, visit our [community resources](https://docs.greatexpectations.io/docs/core/introduction/community_resources#contribute-code-or-documentation).

After you submit your PR, keep the page open and **monitor the statuses of the various checks made by our continuous integration process at the bottom of the page. Please fix any issues that come up** and [reach out on Slack](https://greatexpectations.io/slack) if you need help. Thanks for contributing!

fredwang1012 committed:
Bug fix for Databricks SQL batch replacement

wookasz commented on great_expectations/execution_engine/sqlalchemy_execution_engine.py:
```diff
@@ -328,6 +328,28 @@ def __init__(  # noqa: C901, PLR0912, PLR0913, PLR0915 # FIXME CoP
+                
+                # Replace the dialect's preparer with our custom one
+                self.engine.dialect.identifier_preparer = DatabricksIdentifierPreparer(self.engine.dialect)
+                logger.debug("Successfully patched Databricks dialect to use backticks for identifier quoting")
```
This doesn't seem correct.

The databricks dialect we use already overrides the preparer. See [here](https://github.com/databricks/databricks-sqlalchemy/blob/main/src/databricks/sqlalchemy/base.py#L52).

Furthermore, `identifier_preparer` is a `memoized_property` which I believe is meant to be read-only. It's derived from the dialect's `property` field.

wookasz commented on great_expectations/execution_engine/sqlalchemy_execution_engine.py:
```diff
@@ -1149,7 +1171,33 @@ def _subselectable(self, batch_spec: BatchSpec) -> sqlalchemy.Selectable:
         selectable: sqlalchemy.Selectable
         if table_name:
-            selectable = sa.table(table_name, schema=batch_spec.get("schema_name", None))
+            # Check if we're using Databricks and need special handling
+            if self.dialect_name == GXSqlDialect.DATABRICKS:
+                import re
+                table_name_str = str(table_name)
+                schema_name = batch_spec.get("schema_name", None)
+                schema_name_str = str(schema_name) if schema_name else None
+                
+                # Check if the table name needs special escaping for Databricks
+                if re.match(r"^\d", table_name_str) or re.search(r"[.\s\-#@]", table_name_str):
+                    # For Databricks, create a subquery with backticks
+                    clean_table_name = table_name_str.strip('"').strip("'").strip("`")
+                    if schema_name_str:
+                        clean_schema_name = schema_name_str.strip('"').strip("'").strip("`")
+                        # Use a select with text to control the identifier quoting
+                        selectable = sa.select(sa.text("*")).select_from(
+                            sa.text(f"`{clean_schema_name}`.`{clean_table_name}`")
+                        ).subquery()
+                    else:
+                        selectable = sa.select(sa.text("*")).select_from(
+                            sa.text(f"`{clean_table_name}`")
+                        ).subquery()
+                else:
+                    # Standard table that doesn't need special escaping
+                    selectable = sa.table(table_name, schema=schema_name)
+            else:
+                # Standard behavior for all other databases
+                selectable = sa.table(table_name, schema=batch_spec.get("schema_name", None))
```
Let me get someone who is more familiar with this part of the codebase to take a look at this part.

NathanFarmer replied:
@fredwang1012 can you add some integration test cases to `https://github.com/great-expectations/great_expectations/tree/develop/tests/integration/data_sources_and_expectations/expectations` demonstrating the bug this is fixing? You can find the testing patterns in other files in that directory. Your tests just need to pass with `DatabricksDatasourceTestConfig`.

fredwang1012 replied:
I have updated the PR with a simpler fix for this and added the integration test cases. Please let me know if further changes are needed.

fredwang1012 committed:
Update sqlalchemy_execution_engine.py and add test_databricks_special_table_names.py

fredwang1012 committed:
Cleaned up some changes

NathanFarmer commented on tests/integration/data_sources_and_expectations/expectations/test_databricks_special_table_names.py:
```diff
@@ -0,0 +1,229 @@
+
+
+@pytest.mark.databricks
+class TestDatabricksSpecialTableNames:
```
Please move this entire file into `tests/integration/data_sources_and_expectations/data_sources` now that I see what the issue is.

NathanFarmer commented on great_expectations/execution_engine/sqlalchemy_execution_engine.py:
```diff
@@ -1144,12 +1144,39 @@ def _build_selectable_from_batch_spec(self, batch_spec: BatchSpec) -> sqlalchemy
         selectable: sqlalchemy.Selectable
         if table_name:
-            selectable = sa.table(table_name, schema=batch_spec.get("schema_name", None))
+            # Handle Databricks identifier quoting if needed
+            if self.dialect_name == GXSqlDialect.DATABRICKS:
+                import re
+                schema_name = batch_spec.get("schema_name", None)
+                table_name_str = str(table_name)
+                schema_name_str = str(schema_name) if schema_name else None
+
+                # Check if the table name needs special escaping for Databricks
+                if re.match(r"^\d", table_name_str) or re.search(r"[.\s\-#@]", table_name_str):
+                    # For Databricks, create a subquery with backticks
+                    clean_table_name = table_name_str.strip('"').strip("'").strip("`")
+                    if schema_name_str:
+                        clean_schema_name = schema_name_str.strip('"').strip("'").strip("`")
+                        # Use a select with text to control the identifier quoting
+                        selectable = sa.select(sa.text("*")).select_from(
+                            sa.text(f"`{clean_schema_name}`.`{clean_table_name}`")
+                        ).subquery()
+                    else:
+                        selectable = sa.select(sa.text("*")).select_from(
+                            sa.text(f"`{clean_table_name}`")
+                        ).subquery()
+                else:
+                    # Standard table that doesn't need special escaping
+                    selectable = sa.table(table_name, schema=schema_name)
+            else:
+                # Standard behavior for all other databases
+                selectable = sa.table(table_name, schema=batch_spec.get("schema_name", None))
```
Instead of adding a new condition here, we would rather keep this logic [in this table_name validator](https://github.com/great-expectations/great_expectations/blob/967c2b669960150b3370ba5a176dac0c0493a65e/great_expectations/datasource/fluent/databricks_sql_datasource.py#L141). We already have logic there for handling "quoted" table names. Feel free to rename that method if it makes more sense to (a validator gets called automatically based on the decorator string). Hopefully, it will be easier to manipulate the table name string directly, rather than work with all of these SQL Alchemy constructs.

NathanFarmer reviewed (changes requested):
Thanks for this contribution @fredwang1012! This "quoted identifiers" bug, is one we have been tackling piece by piece for a long time. There are already some existing tests for similar cases in [this file](https://github.com/great-expectations/great_expectations/blob/develop/tests/integration/fluent/test_sql_datasources.py) that may or may not start failing once you move the changes into `DatabricksTableAsset`.

I also noticed that the linter is failing. You can run the linter at any time by running `invoke lint`. Also, if you run `pre-commit install`, the linter will run automatically before each commit.

fredwang1012 committed:
feat: enhance Databricks table name validators to handle special characters

- Enhanced _resolve_quoted_name validator to automatically detect and quote table names with special characters (digits, spaces, hyphens, dots, symbols)
- Added _resolve_quoted_schema_name validator to handle schema names with special characters
- Moved test file to correct directory structure
- Fixes issue with table/schema names starting with digits or containing special characters

fredwang1012 committed:
Update sqlalchemy_execution_engine.py

NathanFarmer commented on great_expectations/datasource/fluent/databricks_sql_datasource.py:
```diff
@@ -139,6 +139,8 @@ class DatabricksTableAsset(SqlTableAsset):
     @pydantic.validator("table_name")
     @override
     def _resolve_quoted_name(cls, table_name: str) -> str | quoted_name:
+        import re
```
This should be imported at the top of the file

NathanFarmer commented on great_expectations/datasource/fluent/databricks_sql_datasource.py:
```diff
@@ -147,18 +149,69 @@ def _resolve_quoted_name(cls, table_name: str) -> str | quoted_name:
+        if schema_name is None:
+            return schema_name
+            
+        import re
```
Once you move this you won't need to import it twice

NathanFarmer commented on great_expectations/datasource/fluent/databricks_sql_datasource.py:
```diff
@@ -147,18 +149,69 @@ def _resolve_quoted_name(cls, table_name: str) -> str | quoted_name:
+        
+        schema_name_is_quoted: bool = cls._is_bracketed_by_quotes(schema_name)
+
+        from great_expectations.compatibility import sqlalchemy
```
Same with this one

NathanFarmer reviewed (changes requested):
This is looking pretty good! Running `pre-commit install` or `invoke lint` will get you past the linting errors so the rest of the tests will run.

fredwang1012 committed:
fix: Handle Databricks table names in Custom SQL Expectations

- Fix QueryMetricProvider to use dialect-specific identifier quoting
- Update _get_substituted_batch_subquery_from_query_and_batch_selectable to handle TableClause objects
- Use dialect's identifier preparer to ensure correct quote characters (backticks for Databricks)
- Add test for Custom SQL Expectations with special table names
- Fixes PARSE_SYNTAX_ERROR when using table names starting with digits in Custom SQL

This completes the fix for Databricks table identifier quoting across all expectation types

fredwang1012 committed:
int

fredwang1012 committed:
cleaned up imports and improved logic

fredwang1012 committed:
Merge branch 'databricks-quote-bug' of https://github.com/fredwang1012/great_expectations into databricks-quote-bug

fredwang1012 committed:
Revert changes to query_metric_provider.py

fredwang1012 committed:
Update sqlalchemy_execution_engine.py

fredwang1012 committed:
Update query_metric_provider.py

fredwang1012 committed:
Update query_metric_provider.py

fredwang1012 committed:
Update query_metric_provider.py

fredwang1012 committed:
Merge branch 'develop' into databricks-quote-bug

fredwang1012 committed:
Fix linting issues: move imports to top of file

fredwang1012 commented:
@NathanFarmer I've addressed the linting issues by moving the imports to the top of the file. Ready for re-review and workflow approval. Thanks!

NathanFarmer commented on great_expectations/datasource/fluent/databricks_sql_datasource.py:
```diff
@@ -139,38 +142,101 @@ class DatabricksTableAsset(SqlTableAsset):
+        ensures the Databricks dialect uses backticks for special identifiers.
+        """
+        # If it's already a quoted_name object, return as-is
+        if hasattr(sqlalchemy, "quoted_name") and hasattr(table_name, "__class__"):  # type: ignore[truthy-function]
```
This type ignore comment needs to be removed

NathanFarmer commented on great_expectations/datasource/fluent/databricks_sql_datasource.py:
```diff
@@ -139,38 +142,101 @@ class DatabricksTableAsset(SqlTableAsset):
+            return schema_name
+
+        # If it's already a quoted_name object, return as-is
+        if hasattr(sqlalchemy, "quoted_name") and hasattr(schema_name, "__class__"):  # type: ignore[truthy-function]
```
This type ignore comment needs to be removed

NathanFarmer reviewed (changes requested):
According to the static analysis staged, the type ignore comments on these lines need to be removed. You can run the type checker by running `invoke types`. Once you get past that stage we should be able to see if the databricks tests pass.

fredwang1012 committed:
fix type comments

fredwang1012 committed:
Merge branch 'develop' into databricks-quote-bug

fredwang1012 committed:
fixed test file

fredwang1012 committed:
Merge branch 'databricks-quote-bug' of https://github.com/fredwang1012/great_expectations into databricks-quote-bug

fredwang1012 committed:
cleaned lint

fredwang1012 committed:
Merge branch 'develop' into databricks-quote-bug

fredwang1012 committed:
fix test script

fredwang1012 committed:
Merge branch 'databricks-quote-bug' of https://github.com/fredwang1012/great_expectations into databricks-quote-bug

fredwang1012 committed:
lint check

fredwang1012 committed:
unit test error fix

fredwang1012 committed:
fixed tests again

fredwang1012 committed:
lint format

fredwang1012 committed:
Merge branch 'develop' into databricks-quote-bug

NathanFarmer commented on tests/integration/data_sources_and_expectations/data_sources/test_databricks_special_table_names.py:
```diff
@@ -0,0 +1,150 @@
+            pytest.skip("Ephemeral backend doesn't support special table names")
+
+        # Skip if no real Databricks connection is available
+        pytest.skip("Integration test requires real Databricks connection")
```
As far as I can tell, this test doesn't assert anything

NathanFarmer commented on tests/integration/data_sources_and_expectations/data_sources/test_databricks_special_table_names.py:
```diff
@@ -0,0 +1,150 @@
+            pytest.skip("Ephemeral backend doesn't support special table names")
+
+        # Skip if no real Databricks connection is available
+        pytest.skip("Integration test requires real Databricks connection")
```
This test does nothing

NathanFarmer commented on tests/integration/data_sources_and_expectations/data_sources/test_databricks_special_table_names.py:
```diff
@@ -0,0 +1,150 @@
+            pytest.skip("Ephemeral backend doesn't support special table names")
+
+        # Skip if no real Databricks connection is available
+        pytest.skip("Integration test requires real Databricks connection")
```
This test does nothing

NathanFarmer commented on tests/integration/data_sources_and_expectations/data_sources/test_databricks_special_table_names.py:
```diff
@@ -0,0 +1,150 @@
+    def test_resolve_quoted_schema_name_none(self):
+        """Test that None schema names remain None."""
+        result = DatabricksTableAsset._resolve_quoted_schema_name(None)
+        assert result is None
```
These are unit tests since they only test the inputs and outputs of this one method (you don't need a databricks connection to test that). They should be marked with `@pytest.mark.unit` and moved into a new file `tests/datasource/fluent/data_asset/databricks_table_asset.py`.

NathanFarmer reviewed (changes requested):
We are getting closer, but we definitely need some end-to-end integration tests. You were pretty close before you completely changed the `tests/integration/data_sources_and_expectations/data_sources/test_databricks_special_table_names.py` file. Commit `8f9ab2b5b59241824afb391353158a229ee49034` was the last time I reviewed and the file was close then, you were just attempting to use fixtures that didn't exist. You should remove the `context` and `databricks_datasource` fixtures from those tests and then use `DatabricksBatchTestSetup` with `DatabricksDatasourceTestConfig`. You should be able to pass the special table names to either of those classes. Follow the established patterns in `tests/integration/data_sources_and_expectations/data_sources/test_postgresql.py` except pass the `table_name`.

fredwang1012 commented:
I mistakenly thought the tests were supposed to be for the changes made in the PR, not the end-to-end integration tests. I will change them back. Sorry about that.

fredwang1012 committed:
updated tests

fredwang1012 committed:
Merge branch 'develop' into databricks-quote-bug

fredwang1012 committed:
Resolve merge conflict - keep new integration tests

fredwang1012 committed:
Merge branch 'develop' into databricks-quote-bug

fredwang1012 committed:
Merge branch 'develop' into databricks-quote-bug

fredwang1012 committed:
Merge branch 'develop' into databricks-quote-bug

fredwang1012 committed:
Merge branch 'develop' into databricks-quote-bug

fredwang1012 committed:
Merge branch 'develop' into databricks-quote-bug

wookasz commented:
Closing for lack of activity. Please don't hesitate to reopen if there are additional changes or if you feel this was done in error.
